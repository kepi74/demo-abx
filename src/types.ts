export interface IPerson {
  email: string;
  id: string;
  name?: string;
  surname: string;
}
