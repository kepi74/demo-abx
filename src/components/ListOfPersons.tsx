import i18next from 'i18next';
import * as React from 'react';
import { Spinner, Table } from 'reactstrap';
import { IPerson } from '../types';

const renderPersonRow = ({ id, name, surname, email }: IPerson) => (
  <tr key={id}>
    <td>{name || ''}</td>
    <td>{surname}</td>
    <td>
      <a href={`mailto:${email}`}>{email}</a>
    </td>
  </tr>
);

interface IListOpPersonsProps {
  t: i18next.TFunction;
  persons?: IPerson[];
  loading: boolean;
  error: boolean;
}

const ListOfPersonsTBody: React.FC<IListOpPersonsProps> = ({
  t,
  persons,
  loading,
  error,
}) => {
  if (loading) {
    return (
      <tr>
        <td colSpan={3}>
          <Spinner color="secondary" size="sm" />
          {' '}
          {t('list.loading')}
        </td>
      </tr>
    );
  }
  if (error || !persons || !persons.length) {
    return (
      <tr>
        <td colSpan={3}>{t('list.noData')}</td>
      </tr>
    );
  }
  return (
    <>
      {persons.map(renderPersonRow)}
    </>
  );
};

const ListOfPersons: React.FC<IListOpPersonsProps> = ({
  t,
  ...rest
}) => {
  return (
    <Table>
      <thead>
        <tr>
          <th>{t('person.name')}</th>
          <th>{t('person.surname')}</th>
          <th>{t('person.email')}</th>
        </tr>
      </thead>
      <tbody>
        <ListOfPersonsTBody t={t} {...rest} />
      </tbody>
    </Table>
  );
};

export default ListOfPersons;
