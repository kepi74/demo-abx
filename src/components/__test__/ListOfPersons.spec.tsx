import i18next from 'i18next';
import * as React from 'react';
import { default as TestRenderer } from 'react-test-renderer';
import { IPerson } from '../../types';
import ListOfPersons from '../ListOfPersons';

describe('ListOfPersons Component', () => {
  const t = (key: string) => key;

  const persons: IPerson[] = [
    { email: 'kepi@kepi.name', id: 'aaa', name: 'Pavel', surname: 'Kepka' },
    { email: 'janca.kantoj@gmail.com', id: 'bbb', name: 'Jana', surname: 'KantorovÃ¡' },
  ];

  it('renders spinner if loading property is true', () => {
    const tree = TestRenderer
      .create(<ListOfPersons t={t as i18next.TFunction} loading={true} error={false} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders noData row if error property is true', () => {
    const tree = TestRenderer
      .create(<ListOfPersons t={t as i18next.TFunction} loading={false} error={true} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders noData row if persons property is empty array', () => {
    const tree = TestRenderer
      .create(
        <ListOfPersons t={t as i18next.TFunction} loading={false} error={false} persons={[]} />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders noData row if persons property is undefined', () => {
    const tree = TestRenderer
      .create(
        <ListOfPersons t={t as i18next.TFunction} loading={false} error={false} />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders content of table if persons property is provided', () => {
    const tree = TestRenderer
      .create(
        <ListOfPersons
          t={t as i18next.TFunction}
          loading={false}
          error={false}
          persons={persons}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
