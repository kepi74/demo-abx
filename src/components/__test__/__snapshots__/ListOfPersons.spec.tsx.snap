// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListOfPersons Component renders content of table if persons property is provide 1`] = `
<table
  className="table"
>
  <thead>
    <tr>
      <th>
        person.name
      </th>
      <th>
        person.surname
      </th>
      <th>
        person.email
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Pavel
      </td>
      <td>
        Kepka
      </td>
      <td>
        <a
          href="mailto:kepi@kepi.name"
        >
          kepi@kepi.name
        </a>
      </td>
    </tr>
    <tr>
      <td>
        Jana
      </td>
      <td>
        Kantorová
      </td>
      <td>
        <a
          href="mailto:janca.kantoj@gmail.com"
        >
          janca.kantoj@gmail.com
        </a>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`ListOfPersons Component renders content of table if persons property is provided 1`] = `
<table
  className="table"
>
  <thead>
    <tr>
      <th>
        person.name
      </th>
      <th>
        person.surname
      </th>
      <th>
        person.email
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Pavel
      </td>
      <td>
        Kepka
      </td>
      <td>
        <a
          href="mailto:kepi@kepi.name"
        >
          kepi@kepi.name
        </a>
      </td>
    </tr>
    <tr>
      <td>
        Jana
      </td>
      <td>
        Kantorová
      </td>
      <td>
        <a
          href="mailto:janca.kantoj@gmail.com"
        >
          janca.kantoj@gmail.com
        </a>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`ListOfPersons Component renders noData row if error property is true 1`] = `
<table
  className="table"
>
  <thead>
    <tr>
      <th>
        person.name
      </th>
      <th>
        person.surname
      </th>
      <th>
        person.email
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        colSpan={3}
      >
        list.noData
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`ListOfPersons Component renders noData row if persons property is empty array 1`] = `
<table
  className="table"
>
  <thead>
    <tr>
      <th>
        person.name
      </th>
      <th>
        person.surname
      </th>
      <th>
        person.email
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        colSpan={3}
      >
        list.noData
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`ListOfPersons Component renders noData row if persons property is undefined 1`] = `
<table
  className="table"
>
  <thead>
    <tr>
      <th>
        person.name
      </th>
      <th>
        person.surname
      </th>
      <th>
        person.email
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        colSpan={3}
      >
        list.noData
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`ListOfPersons Component renders spinner if loading property is true 1`] = `
<table
  className="table"
>
  <thead>
    <tr>
      <th>
        person.name
      </th>
      <th>
        person.surname
      </th>
      <th>
        person.email
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        colSpan={3}
      >
        <div
          className="spinner-border-sm spinner-border text-secondary"
          role="status"
        >
          <span
            className="sr-only"
          >
            Loading...
          </span>
        </div>
         
        list.loading
      </td>
    </tr>
  </tbody>
</table>
`;
